src/OpenSpaceToolkit/Astrodynamics/Trajectory/Orbit/Model/Kepler/COE.cpp

```diff
@@ -641,6 +642,159 @@ COE COE::FromSIVector(const Vector6d& aCOEVector, const AnomalyType& anAnomalyTy
+    const Array<Angle> criticalAops = {Angle::Degrees(90.0), Angle::Degrees(270.0)};
+
+    const Real equatorialRadius_meters = anEquatorialRadius.inMeters();
+    const Real j2 = aJ2;
+    const Real j3 = aJ3;
+
+    // ecc =~ eccCoefficient * sin(incl)
+    const Real eccCoefficient = -j3 * equatorialRadius_meters / 2.0 / j2 / aSemiMajorAxis.inMeters();
+
+    const auto eccentricityFromInclination = [&eccCoefficient](const Angle& inclination) -> Real
+    {
+        return eccCoefficient * std::sin(inclination.inRadians());
+    };
+
+    const auto inclinationFromEccentricity = [&eccCoefficient](const Real& eccentricity) -> Angle
+    {
+        return Angle::Radians(asin(eccentricity / eccCoefficient));
+    };
+
+    // TBI: this only works because neither set of critical angles are 0/360 degrees
+    const auto isCritical = [](const Angle& angle, const Array<Angle>& criticalAngles) -> bool
+    {
+        const Real epsilon = Angle::Arcseconds(1.0).inRadians();  // TBI: make configurable?
+
+        return (angle.inRadians().isNear(criticalAngles[0].inRadians(), epsilon)) ||
+               (angle.inRadians().isNear(criticalAngles[1].inRadians(), epsilon));
+    };
+
+    // Use the provided AoP, or default to a critical value
+    const Angle aop = anAop.isDefined() ? anAop : criticalAops[0];
+
+    // If AoP matches a critical value
+    if (isCritical(aop, criticalAops))
+    {
+        Angle inclination = Angle::Undefined();
+        Real eccentricity = Real::Undefined();
+
+        // If ecc not defined, use the given inclination or default to critical inclination, and calculate ecc
+        if (!anEccentricity.isDefined())
+        {
+            inclination = anInclination.isDefined() ? anInclination : criticalInclinations[0];
+            eccentricity = eccentricityFromInclination(inclination);
+        }
+        // If ecc defined, use it to calculate inc
+        else
+        {
+            // If the eccentricity is larger than this value, the approximation isn't valid
+            if (anEccentricity > eccCoefficient)
+            {
+                throw ostk::core::error::RuntimeError(
+                    "Provided eccentricity [" + anEccentricity.toString() + "] cannot be greater than " +
+                    eccCoefficient.toString()
+                );
+            }
+            inclination = inclinationFromEccentricity(anEccentricity);
+            eccentricity = anEccentricity;
+        }
+
+        return {
+            aSemiMajorAxis,
+            eccentricity,
+            inclination,
+            aRaan,
+            aop,
+            aTrueAnomaly,
+        };
+    }
+
+    // If AoP is (given) and not a critical angle, then the inclination must be critical
+    const Angle inclination = anInclination.isDefined() ? anInclination : criticalInclinations[0];
+    if (!isCritical(inclination, criticalInclinations))
+    {
+        throw ostk::core::error::RuntimeError("Provided inclination must be a critical value");
```

Comment:
so if a user intended to provide a critical inclination and AoP - but was slightly off, they'll only get an error message saying "Provided inclination must be a critical value" - even though it wasn't necessarily just the inclination that was incorrectly non-critical?
Is that a correct interpretation of the code? If so I propose shifting around the logic / error message to yield a more accurate portrayal of what the user did incorrectly